### PR TITLE
Checkbox: indeterminate state

### PR DESCRIPTION
**Pull request by maxwellred**
#### Pull request checklist
- [] Addresses issue #5390 
- [x] Include a change request file using `$ npm run change`

#### Description of changes

Added an indeterminate visual state to Checkbox. Users are able to activate this state by passing in "indeterminate" in props.

#### Screenshots

Screen Shot 2019-07-15 at 5 52 07 PM

###### Microsoft Reviewers: [Open in CodeFlow](http://wpcp.azurewebsites.net/CodeFlowProtocolProxy2.php?pullrequest=https://github.com/OfficeDev/office-ui-fabric-react/pull/9815)

### CODE REVIEW

File: packages/office-ui-fabric-react/src/components/Checkbox/Checkbox.base.tsx [hunk 1]
```diff
@@ -7,6 +7,10 @@ import { KeytipData } from '../../KeytipData';
 export interface ICheckboxState {
   /** Is true when Uncontrolled control is checked. */
   isChecked?: boolean;
+  isIndeterminate?: boolean;
+  /** Since indeterminant props are always truthy,
+   * isIndeterCheck state acts as a secondary logic gate to turn off indeterminate state after the first click */
+  isIndeterCheck?: boolean;
```
Comment (Vitalius1) [inline packages/office-ui-fabric-react/src/components/Checkbox/Checkbox.base.tsx]: Very confused why you would need this prop?

File: packages/office-ui-fabric-react/src/components/Checkbox/Checkbox.base.tsx [hunk 2]
```diff
@@ -7,6 +7,10 @@ import { KeytipData } from '../../KeytipData';
 export interface ICheckboxState {
   /** Is true when Uncontrolled control is checked. */
   isChecked?: boolean;
+  isIndeterminate?: boolean;
```
Comment (Vitalius1) [inline packages/office-ui-fabric-react/src/components/Checkbox/Checkbox.base.tsx]: Why not match the name of the prop?
What happens if both `indeterminate` and `checked` props are enabled? Perhaps needs some details in documentation about that `indeterminate` wins priority initially and then it’s never updated even if they try. This will remove the need of the second state.

File: packages/office-ui-fabric-react/etc/office-ui-fabric-react.api.md [hunk 3]
```diff
@@ -7797,7 +7803,7 @@ export class LayerBase extends React.Component<ILayerProps, ILayerBaseState> {
     }
 
 // Warning: (ae-forgotten-export) The symbol "ILayerHostProps" needs to be exported by the entry point index.d.ts
-//
+//
```
Comment (Vitalius1) [inline packages/office-ui-fabric-react/etc/office-ui-fabric-react.api.md]: I wonder why this got updated with an empty space.

**Comment by cliffkoh** [comment]
It looks like this is missing a changefile...

File: apps/vr-tests/src/stories/Checkbox.stories.tsx [hunk 4]
```diff
@@ -23,6 +23,10 @@ storiesOf('Checkbox', module)
   .addStory('Checked disabled', () => (
     <Checkbox label="Checked disabled checkbox" checked disabled />
   ))
+  .addStory('Indeterminate', () => <Checkbox label="Indeterminate checkbox" indeterminate />)
```
Comment (cliffkoh) [inline apps/vr-tests/src/stories/Checkbox.stories.tsx]: :+1:

File: packages/office-ui-fabric-react/src/components/Checkbox/Checkbox.styles.ts [hunk 5]
```diff
@@ -25,18 +25,49 @@ export const getStyles = (props: ICheckboxStyleProps): ICheckboxStyles => {
...
+  const checkboxBorderIndeterminateColor = semanticColors.inputBackgroundChecked;
   const checkboxBorderColorChecked = semanticColors.inputBackgroundChecked;
   const checkboxBorderColorDisabled = semanticColors.disabledBodySubtext;
   const checkboxBorderHoveredColor = semanticColors.inputBorderHovered;
+  const checkboxBorderIndeterminateHoveredColor = semanticColors.inputBackgroundCheckedHovered;
   const checkboxBackgroundChecked = semanticColors.inputBackgroundChecked;
   // TODO: after updating the semanticColors slots mapping following 2 tokens need to be semanticColors.inputBackgroundCheckedHovered
-  const checkboxBackgroundCheckedHovered = palette.themeDark;
-  const checkboxBorderColorCheckedHovered = palette.themeDark;
+  const checkboxBackgroundCheckedHovered = semanticColors.inputBackgroundCheckedHovered;
```
Comment (cliffkoh) [inline packages/office-ui-fabric-react/src/components/Checkbox/Checkbox.styles.ts]: :+1:

File: packages/office-ui-fabric-react/src/components/Checkbox/Checkbox.base.tsx (hunk 1 above)
Comment (maxwellred) [reply packages/office-ui-fabric-react/src/components/Checkbox/Checkbox.base.tsx]: It was to help the internal logic differentiate between the first click (to remove indeterminate state) versus successive clicks but I have since changed the logic to only work with this.state.isIndeterminate.

File: packages/office-ui-fabric-react/src/components/Checkbox/Checkbox.base.tsx (hunk 2 above)
Comment (maxwellred) [reply packages/office-ui-fabric-react/src/components/Checkbox/Checkbox.base.tsx]: I was just following suit with how the previous developer defined isChecked state as well as making it potentially easier to differentiate between props and state. I have currently reworked logic to make indeterminate state precede checked and defaultChecked, thanks!

**Comment by maxwellred** [comment]
Hi @cliffkoh , good point on the third state of the checkbox. I have changed the isChecked base component state to be set to "undefined" when indeterminate props are passed in. Thanks!

**Comment by cliffkoh** [comment]
In this case I would expect `indeterminate` should only enable the third state but not set it.. This will require quite a bit of other changes, including to screener tests..

Also tagging @dzearing in case he has any feedback on the API surface since he has spent way more mental cycles than me thinking about such things (and this is quite a significant context switch for me atm).

File: common/changes/office-ui-fabric-react/v-mare-Checkbox-indeterminate-state_2019-07-16-00-47.json [hunk 6]
```diff
@@ -0,0 +1,11 @@
+{
+  "changes": [
+    {
```
Comment (Vitalius1) [inline common/changes/office-ui-fabric-react/v-mare-Checkbox-indeterminate-state_2019-07-16-00-47.json]: This file needs to be removed from the PR in favor of the new format.

File: change/office-ui-fabric-react-2019-07-16-17-27-21-v-mare-Checkbox-indeterminate-state.json [hunk 7]
```diff
@@ -0,0 +1,8 @@
+{
+  "comment": "Updated API documentation",
```
Comment (Vitalius1) [inline change/office-ui-fabric-react-2019-07-16-17-27-21-v-mare-Checkbox-indeterminate-state.json]: Change the comment in with the one from the change file that needs to be removed.

File: packages/office-ui-fabric-react/src/components/Checkbox/Checkbox.types.ts [hunk 8]
```diff
@@ -122,6 +122,12 @@ export interface ICheckboxProps extends React.ButtonHTMLAttributes<HTMLElement |
    * Optional keytip for this checkbox
    */
   keytipProps?: IKeytipProps;
+
+  /**
+   * Optional indeterminate state for this checkbox.
+   * Setting indeterminate state takes precedence over checked and defaultChecked props given.
+   */
+  indeterminate?: boolean;
```
Comment (Vitalius1) [inline packages/office-ui-fabric-react/src/components/Checkbox/Checkbox.types.ts]: Do we allow to control this prop? Meaning that can we toggle on from the consuming app the indeterminate state after the first click? I would assume not. If so, worth mentioning that in the docs.

Comment (maxwellred) [reply packages/office-ui-fabric-react/src/components/Checkbox/Checkbox.types.ts]: Correct, It is applied to the checkbox on load but then disappears after the first click without reappearing under any circumstance. Thanks, I'll add that to the docs!

File: packages/office-ui-fabric-react/src/components/Checkbox/Checkbox.base.tsx [hunk 9]
```diff
@@ -21,10 +22,22 @@ export class CheckboxBase extends React.Component<ICheckboxProps, ICheckboxState
...
 
   public static getDerivedStateFromProps(props: ICheckboxProps, state: ICheckboxState): ICheckboxState {
-    if (props.checked !== undefined) {
+    if (!props.indeterminate) {
+      if (props.checked) {
+        return {
+          ...state,
+          isIndeterminate: false,
+          isChecked: !!props.checked
+        };
```
Comment (cliffkoh) [inline packages/office-ui-fabric-react/src/components/Checkbox/Checkbox.base.tsx]: These few lines differ from your API types documentation: 
> * Setting indeterminate state takes precedence over checked and defaultChecked props given.

File: packages/office-ui-fabric-react/src/components/Checkbox/Checkbox.base.tsx [hunk 10]
```diff
@@ -155,16 +170,20 @@ export class CheckboxBase extends React.Component<ICheckboxProps, ICheckboxState
...
-        this.setState({ isChecked: !isChecked });
+      if (isIndeterminate) {
+        if (onChange) {
+          onChange(ev, !isChecked);
+        }
+        this.setState({ isIndeterminate: false });
+      } else {
+        if (onChange) {
+          onChange(ev, !isChecked);
+        }
```
Comment (cliffkoh) [inline packages/office-ui-fabric-react/src/components/Checkbox/Checkbox.base.tsx]: looks like line 182 - 184 are the same in both branches of the if clause .. could you simplify it to something like
```
if(onChange){ 
  onChange(ev, !isChecked)
}

if(isIndeterminate){
  this.setState({ isIndeterminate: false });
}
```

**Comment by cliffkoh** [comment]
>    */

Please also update the relevant pieces to document that checked will be undefined if the component is indeterminate

---
Refers to: packages/office-ui-fabric-react/src/components/Checkbox/Checkbox.types.ts:59 in f51e3f0. [](commit_id = f51e3f0f29e7286e29bff94878b437b45bf9a736, deletion_comment = False)

File: packages/office-ui-fabric-react/src/components/Checkbox/Checkbox.types.ts (hunk 8 above)
Comment (dzearing) [reply packages/office-ui-fabric-react/src/components/Checkbox/Checkbox.types.ts]: This seems consistent with other apis. All of them use `indeterminate` as an overridden value. Semantic uses `defaultIndeterminate` to support uncontrolled modes, but it is the only one here afaik.

https://material-ui.com/api/checkbox
https://react.semantic-ui.com/modules/checkbox/#states-indeterminate
https://ant.design/components/checkbox/
https://github.com/carbon-design-system/carbon-components-react/blob/master/src/components/Checkbox/Checkbox.js

Comment (dzearing) [reply packages/office-ui-fabric-react/src/components/Checkbox/Checkbox.types.ts]: Wait. I might have read your comment wrong @maxwellred . I think if you have `indeterminate` flag on, and you click on the checkbox, it stays in an indeterminate state until the caller reacts to `onChange` and removes it.

If you want an uncontrolled behavior, I suggest `defaultIndeterminate` to be symmetrical with `defaultChecked`.

Comment (maxwellred) [reply packages/office-ui-fabric-react/src/components/Checkbox/Checkbox.types.ts]: Hi @dzearing i've set up indeterminate state to work as you describe where the dev passes in the indeterminate prop which sets isChecked to undefined and keeps isIndeterminate state set to true until onChange (per successive click) removes it. 

I'm a little confused as to how an uncontrolled behavior is to be implemented as the developer has to manually place the indeterminate props on components. Is the intention that we place a native indeterminate attribute on the Checkbox's <input> tag?

Comment (cliffkoh) [reply packages/office-ui-fabric-react/src/components/Checkbox/Checkbox.types.ts]: @maxwellred I agree with @dzearing here. Atm your `indeterminate` prop, per its JSDoc, is supposed to behave like an uncontrolled prop (i.e. what `defaultIndeterminate` would have done) except that at the same time, it behaves like a controlled prop (because you update state in `getDerivedStateFromProps`).

If you could add in an example in the documentation state for using indeterminate checkbox in a controlled fashion, you should be able to see what the problem is (which is that as you control the checked/unchecked state of the component, the checkbox keeps flipping (or staying) in indeterminate .. violating what is implied in the JSDoc about how this property works.

It is probably cleaner and less confusing if we instead have dedicated `indeterminate` and `defaultIndeterminate` (mirroring checked and defaultChecked)  prop - if you want to use checkbox in a controlled fashion, use "checked" and "indeterminate" together. Otherwise, use defaultIndeterminate + defaultChecked.

Comment (maxwellred) [reply packages/office-ui-fabric-react/src/components/Checkbox/Checkbox.types.ts]: @cliffkoh @dzearing Alright I updated the functionality to mirror checked/defaultChecked, documentation, and the aria-checked attribute. I changed the functionality of the indeterminate state to be only visual and not affect checked state. My assumption was that since indeterminate state is an on load operation that the checkbox ought to be in this undefined state however the spec goes into how upon removing indeterminate state the checkbox should return the previous state it was set on (my bad).

File: packages/office-ui-fabric-react/src/components/Checkbox/Checkbox.base.tsx [hunk 11]
```diff
@@ -21,11 +22,18 @@ export class CheckboxBase extends React.Component<ICheckboxProps, ICheckboxState
...
 
   public static getDerivedStateFromProps(props: ICheckboxProps, state: ICheckboxState): ICheckboxState {
-    if (props.checked !== undefined) {
+    if (props.indeterminate !== undefined && state.isIndeterminate) {
       return {
         ...state,
-        isChecked: !!props.checked
+        isIndeterminate: !!props.indeterminate
       };
+    } else {
```
Comment (Vitalius1) [inline packages/office-ui-fabric-react/src/components/Checkbox/Checkbox.base.tsx]: because you are returning in the first if statement you can drop the `else` and just do another if statement right after.

File: packages/office-ui-fabric-react/src/components/Checkbox/Checkbox.types.ts [hunk 12]
```diff
@@ -122,6 +122,32 @@ export interface ICheckboxProps extends React.ButtonHTMLAttributes<HTMLElement |
    * Optional keytip for this checkbox
    */
   keytipProps?: IKeytipProps;
+
+  /**
+   * Optional controlled indeterminate visual state for checkbox. Use this if you want to control
+   * the indeterminate state. Setting indeterminate state takes visual precedence
+   * over checked or defaultChecked props given but does not affect checked state.
+   * This is should not be a toggleable state.
```
Comment (Vitalius1) [inline packages/office-ui-fabric-react/src/components/Checkbox/Checkbox.types.ts]: `This is should…` 🤔

Comment (maxwellred) [reply packages/office-ui-fabric-react/src/components/Checkbox/Checkbox.types.ts]: I was wondering if it should be "should" since the user is passing their own removeIndeterminate callback handling how indeterminate is removed. As in, if they wanted to handle indeterminate differently than what is established in spec then they could. Or should we push them towards complying with spec?

Comment (Vitalius1) [reply packages/office-ui-fabric-react/src/components/Checkbox/Checkbox.types.ts]: I would say if the user has the ability to control it multiple times during the life of the component it needs to say `should` as it sounds more as a recommendation.

Comment (maxwellred) [reply packages/office-ui-fabric-react/src/components/Checkbox/Checkbox.types.ts]: Sounds good, I'll push up the changes, thanks!

**Comment by micahgodbolt** [comment]
Are we pretty close on this PR? Anytime I can do to help move this forward?

File: packages/office-ui-fabric-react/src/components/Checkbox/Checkbox.types.ts [hunk 13]
```diff
@@ -122,6 +122,32 @@ export interface ICheckboxProps extends React.ButtonHTMLAttributes<HTMLElement |
    * Optional keytip for this checkbox
    */
   keytipProps?: IKeytipProps;
+
+  /**
+   * Optional controlled indeterminate visual state for checkbox. Use this if you want to control
+   * indeterminate state. Setting indeterminate state takes visual precedence
+   * over checked or defaultChecked props given but does not affect checked state.
+   * This is should not be a toggleable state. On load the checkbox will receive indeterminate visual state
```
Comment (cliffkoh) [inline packages/office-ui-fabric-react/src/components/Checkbox/Checkbox.types.ts]: >is should not be [](start = 10, length = 16)

there's some typo with the sentence structure here

File: packages/office-ui-fabric-react/src/components/Checkbox/Checkbox.types.ts [hunk 14]
```diff
@@ -122,6 +122,32 @@ export interface ICheckboxProps extends React.ButtonHTMLAttributes<HTMLElement |
...
+   * and after the first user click it should be removed by your supplied removeIndeterminate callback
```
Comment (cliffkoh) [inline packages/office-ui-fabric-react/src/components/Checkbox/Checkbox.types.ts]: >click [](start = 30, length = 5)

clicks, not click

File: packages/office-ui-fabric-react/src/components/Checkbox/Checkbox.base.tsx [hunk 15]
```diff
@@ -21,12 +22,19 @@ export class CheckboxBase extends React.Component<ICheckboxProps, ICheckboxState
   private _classNames: { [key in keyof ICheckboxStyles]: string };
 
   public static getDerivedStateFromProps(props: ICheckboxProps, state: ICheckboxState): ICheckboxState {
+    if (!props.defaultIndeterminate && state.isIndeterminate) {
+      return {
+        ...state,
```
Comment (cliffkoh) [inline packages/office-ui-fabric-react/src/components/Checkbox/Checkbox.base.tsx]: >...state, [](start = 8, length = 9)

I actually don't know for sure but is this spread really necessary? I do recognize that you probably followed the existing pattern in lines 32-36 though..

File: packages/office-ui-fabric-react/src/components/Checkbox/Checkbox.base.tsx [hunk 16]
```diff
@@ -114,6 +127,7 @@ export class CheckboxBase extends React.Component<ICheckboxProps, ICheckboxState
               aria-describedby={mergeAriaAttributeValues(ariaDescribedBy, keytipAttributes['aria-describedby'])}
               aria-posinset={ariaPositionInSet}
               aria-setsize={ariaSetSize}
+              aria-checked={isIndeterminate ? 'mixed' : isChecked ? 'true' : 'false'}
```
Comment (cliffkoh) [inline packages/office-ui-fabric-react/src/components/Checkbox/Checkbox.base.tsx]: >? 'mixed' [](start = 44, length = 9)

:+1:

File: packages/office-ui-fabric-react/src/components/Checkbox/Checkbox.types.ts [hunk 17]
```diff
@@ -122,6 +122,32 @@ export interface ICheckboxProps extends React.ButtonHTMLAttributes<HTMLElement |
...
+   * This is not a toggleable state. On load the checkbox will receive indeterminate visual state
+   * and after the first user click it will be removed exposing the true state of the checkbox.
+   */
+  defaultIndeterminate?: boolean;
+
+  /**
+   * User provided callback that is called to remove indeterminate visual state when opting to
+   * use the controlled indeterminate prop.
+   */
+  removeIndeterminate?: () => void;
```
Comment (cliffkoh) [inline packages/office-ui-fabric-react/src/components/Checkbox/Checkbox.types.ts]: >removeIndeterminate [](start = 2, length = 19)

should this be `onRemoveIndeterminate()` to be consistent with general callbacks naming convention?

File: packages/office-ui-fabric-react/src/components/Checkbox/Checkbox.base.tsx (hunk 15 above)
Comment (cliffkoh) [reply packages/office-ui-fabric-react/src/components/Checkbox/Checkbox.base.tsx]: @KevinTCoughlin  looked into this and validated it isn't necessary

https://github.com/facebook/react/blob/d9b4c55d53c67045c38b12f5aad57a163e7f63a2/packages/react-reconciler/src/ReactFiberClassComponent.js#L163

File: packages/office-ui-fabric-react/src/components/Checkbox/Checkbox.types.ts (hunk 14 above)
Comment (maxwellred) [reply packages/office-ui-fabric-react/src/components/Checkbox/Checkbox.types.ts]: Would it be altogether better to write "...user's first click..."?

File: packages/office-ui-fabric-react/src/components/Checkbox/Checkbox.base.tsx [hunk 18]
```diff
@@ -21,12 +22,19 @@ export class CheckboxBase extends React.Component<ICheckboxProps, ICheckboxState
...
   public static getDerivedStateFromProps(props: ICheckboxProps, state: ICheckboxState): ICheckboxState {
+    if (!props.defaultIndeterminate && state.isIndeterminate) {
+      return {
+        ...state,
+        isIndeterminate: !!props.indeterminate
+      };
+    }
     if (props.checked !== undefined) {
       return {
         ...state,
```
Comment (ecraig12345) [inline packages/office-ui-fabric-react/src/components/Checkbox/Checkbox.base.tsx]: You can remove this spread too. And the last line of the method can return null.

File: packages/office-ui-fabric-react/src/components/Checkbox/Checkbox.base.tsx [hunk 19]
```diff
@@ -154,16 +168,23 @@ export class CheckboxBase extends React.Component<ICheckboxProps, ICheckboxState
   };
 
   private _onChange = (ev: React.FormEvent<HTMLElement>): void => {
-    const { disabled, onChange } = this.props;
-    const { isChecked } = this.state;
+    const { disabled, onChange, removeIndeterminate, indeterminate, defaultIndeterminate } = this.props;
```
Comment (ecraig12345) [inline packages/office-ui-fabric-react/src/components/Checkbox/Checkbox.base.tsx]: Not the right line for this, but you should add an `indeterminate: boolean` property to `ICheckbox`, and a corresponding getter in class (mirroring `public get checked`).

File: packages/office-ui-fabric-react/src/components/Checkbox/Checkbox.types.ts (hunk 14 above)
Comment (cliffkoh) [reply packages/office-ui-fabric-react/src/components/Checkbox/Checkbox.types.ts]: OH, I misinterpreted :) Sorry :$ 

Yeah I think user's first click would be better :)

**Comment by cliffkoh** [comment]
Could you add a `warnMutuallyExclusive` check to the `indeterminate` and `defaultIndeterminate` props please?

File: packages/office-ui-fabric-react/src/components/Checkbox/Checkbox.base.tsx [hunk 20]
```diff
@@ -154,16 +168,23 @@ export class CheckboxBase extends React.Component<ICheckboxProps, ICheckboxState
...
+          onChange(ev, !isChecked);
+        }
+        if (this.props.checked === undefined) {
+          this.setState({ isChecked: !isChecked });
+        }
+      } else {
+        if (defaultIndeterminate) {
+          this.setState({ isIndeterminate: false });
+        } else if (removeIndeterminate && indeterminate) {
+          removeIndeterminate();
```
Comment (ecraig12345) [inline packages/office-ui-fabric-react/src/components/Checkbox/Checkbox.base.tsx]: `removeIndeterminate` should be called regardless of whether it's controlled or not, to match the pattern with `onChange`

**Comment by ecraig12345** [comment]
> Could you add a `warnMutuallyExclusive` check to the `indeterminate` and `defaultIndeterminate` props please?

Just to be clear, you can combine it into the object passed to the existing `warnMutuallyExclusive` check.

File: packages/office-ui-fabric-react/src/components/Checkbox/Checkbox.base.tsx (hunk 15 above)
Comment (maxwellred) [reply packages/office-ui-fabric-react/src/components/Checkbox/Checkbox.base.tsx]: True, I was following suit with the existing pattern but after removing it and checking it now it works so i'll drop both in both scenarios.

File: packages/office-ui-fabric-react/src/components/Checkbox/Checkbox.base.tsx (hunk 20 above)
Comment (maxwellred) [reply packages/office-ui-fabric-react/src/components/Checkbox/Checkbox.base.tsx]: Sounds good, switched it over to match `onChange`

File: packages/office-ui-fabric-react/src/components/Checkbox/Checkbox.base.tsx (hunk 19 above)
Comment (cliffkoh) [reply packages/office-ui-fabric-react/src/components/Checkbox/Checkbox.base.tsx]: Very good catch :clap:

**Comment by maxwellred** [comment]
@cliffkoh Absolutely! It's been a great experience working with you all. Thanks for taking the time to give so much good feedback!

File: packages/office-ui-fabric-react/src/components/Checkbox/Checkbox.types.ts [hunk 21]
```diff
@@ -122,6 +125,32 @@ export interface ICheckboxProps extends React.ButtonHTMLAttributes<HTMLElement |
...
+   * This is not a toggleable state. On load the checkbox will receive indeterminate visual state
+   * and after the user's first click it will be removed exposing the true state of the checkbox.
+   */
+  defaultIndeterminate?: boolean;
+
+  /**
+   * User provided callback that is called to remove indeterminate visual state when opting to
+   * use the controlled indeterminate prop.
+   */
+  onRemoveIndeterminate?: () => void;
```
Comment (dzearing) [inline packages/office-ui-fabric-react/src/components/Checkbox/Checkbox.types.ts]: Is this needed? I believe `onChange` would be good enough. That is, in a controlled setup, the user would click, which would then call `onChange(ev, true)`. The use would then pipe that into `value`.

**Comment by cliffkoh** [comment]
@msft-github-bot merge this after @ecraig12345 or @dzearing approves

File: packages/office-ui-fabric-react/src/components/Checkbox/Checkbox.types.ts (hunk 21 above)
Comment (maxwellred) [reply packages/office-ui-fabric-react/src/components/Checkbox/Checkbox.types.ts]: Changed it with Elizabeth to where onChange has a third optional parameter to handle indeterminate changes. Thanks!

File: packages/office-ui-fabric-react/src/components/Checkbox/Checkbox.base.tsx [hunk 22]
```diff
@@ -21,13 +22,17 @@ export class CheckboxBase extends React.Component<ICheckboxProps, ICheckboxState
...
+      };
+    }
     if (props.checked !== undefined) {
       return {
-        ...state,
         isChecked: !!props.checked
       };
     }
-    return state;
+    return {};
```
Comment (ecraig12345) [inline packages/office-ui-fabric-react/src/components/Checkbox/Checkbox.base.tsx]: We should return null here--it might still trigger a state update (when it's not actually needed) if you use an empty object.

Comment (ecraig12345) [reply packages/office-ui-fabric-react/src/components/Checkbox/Checkbox.base.tsx]: Just update the return type of the method to include `| null`

File: packages/office-ui-fabric-react/src/components/Checkbox/Checkbox.base.tsx [hunk 23]
```diff
@@ -42,13 +47,15 @@ export class CheckboxBase extends React.Component<ICheckboxProps, ICheckboxState
...
+        checked: 'defaultChecked',
+        indeterminate: 'defaultIndeterminate'
       });
     }
 
     this._id = this.props.id || getId('checkbox-');
     this.state = {
-      isChecked: !!(props.checked !== undefined ? props.checked : props.defaultChecked)
+      isChecked: !!(props.checked !== undefined ? props.checked : props.defaultChecked),
+      isIndeterminate: !!(props.defaultIndeterminate ? props.defaultIndeterminate : props.indeterminate)
```
Comment (ecraig12345) [inline packages/office-ui-fabric-react/src/components/Checkbox/Checkbox.base.tsx]: These checks should be similar to the ones for checked: `!!(props.indeterminate !== undefined ? props.indeterminate : props.defaultIndeterminate)`. The controlled prop is supposed to take precedence over the uncontrolled one.

File: packages/office-ui-fabric-react/src/components/Checkbox/Checkbox.types.ts [hunk 24]
```diff
@@ -56,8 +59,10 @@ export interface ICheckboxProps extends React.ButtonHTMLAttributes<HTMLElement |
 
   /**
    * Callback that is called when the checked value has changed.
+   * However, if the Checkbox has a controlled indeterminate state,
+   * this is called when the indeterminate value has changed.
    */
-  onChange?: (ev?: React.FormEvent<HTMLElement | HTMLInputElement>, checked?: boolean) => void;
+  onChange?: (ev?: React.FormEvent<HTMLElement | HTMLInputElement>, checked?: boolean, wasIndeterminate?: boolean) => void;
```
Comment (dzearing) [inline packages/office-ui-fabric-react/src/components/Checkbox/Checkbox.types.ts]: Why is this needed? I do not believe there is a scenario where this matters.

File: packages/office-ui-fabric-react/src/components/Checkbox/Checkbox.types.ts (hunk 21 above)
Comment (dzearing) [reply packages/office-ui-fabric-react/src/components/Checkbox/Checkbox.types.ts]: I don't see why adding a 3rd value is needed. What is the reasoning

File: packages/office-ui-fabric-react/src/components/Checkbox/Checkbox.types.ts (hunk 24 above)
Comment (maxwellred) [reply packages/office-ui-fabric-react/src/components/Checkbox/Checkbox.types.ts]: @dzearing Per our conversation I reverted onChange functionality back to only handling component checked state. I changed the example file to handle user controlled indeterminate state. Please let me know if there's anymore I can do and thanks!

File: packages/office-ui-fabric-react/src/components/Checkbox/Checkbox.base.tsx [hunk 25]
```diff
@@ -20,14 +21,18 @@ export class CheckboxBase extends React.Component<ICheckboxProps, ICheckboxState
   private _id: string;
   private _classNames: { [key in keyof ICheckboxStyles]: string };
 
-  public static getDerivedStateFromProps(props: ICheckboxProps, state: ICheckboxState): ICheckboxState {
+  public static getDerivedStateFromProps(props: ICheckboxProps, state: ICheckboxState): ICheckboxState | null {
+    if (!props.defaultIndeterminate && state.isIndeterminate) {
```
Comment (dzearing) [inline packages/office-ui-fabric-react/src/components/Checkbox/Checkbox.base.tsx]: You're checking falsey values here. Shouldn't you be checking for an `undefined` exact match?

```tsx
if (props.indeterminate !== undefined) {
  return { isIndeterminate: props.indeterminate };
}
```

Comment (dzearing) [reply packages/office-ui-fabric-react/src/components/Checkbox/Checkbox.base.tsx]: also, it's ignoring `props.checked` in this codepath...

Perhaps a better approach would be:

```tsx
const { 
  indeterminate = state.indeterminate,
  checked = state.checked
 } = props;

return {
  checked,
  indeterminate
};
```

Comment (dzearing) [reply packages/office-ui-fabric-react/src/components/Checkbox/Checkbox.base.tsx]: This way, if props contains a controlled value for indeterminate or checked, it's always preferred over whatever is currently in the state.

Comment (dzearing) [reply packages/office-ui-fabric-react/src/components/Checkbox/Checkbox.base.tsx]: I think there's a confusing scenario where both indeterminate and checked are provided, which seems conflicting.

**Comment by dzearing** [comment]
I don't see any modifications to `Checkbox.test.tsx`. I'd love to see unit tests covering:

1. defaultIndeterminate defaults to indeterminate, and you can click to transition to checked, and onChange files.
2. indeterminate defaults to render indeterminate, and clicking causes an onchange event (with value === true), which if nothing is done, makes the Checkbox stay indeterminate (controlled scenario.)
3. manually transitioning indeterminate from true to false reflects the current checked state.

File: packages/office-ui-fabric-react/src/components/Checkbox/__snapshots__/Checkbox.test.tsx.snap [hunk 26]
```diff
@@ -94,11 +95,25 @@ exports[`Checkbox renders Checkbox correctly 1`] = `
             justify-content: center;
             margin-right: 4px;
             overflow: hidden;
+            position: relative;
             transition-duration: 200ms;
             transition-property: background, border, border-color;
             transition-timing-function: cubic-bezier(.4, 0, .23, 1);
             width: 20px;
           }
+          &:after {
```
Comment (dzearing) [inline packages/office-ui-fabric-react/src/components/Checkbox/__snapshots__/Checkbox.test.tsx.snap]: it seems weird that we're adding this rule without any modifications. Can we only define the after rule when we need it?

File: packages/office-ui-fabric-react/src/components/Checkbox/Checkbox.styles.ts [hunk 27]
```diff
@@ -177,7 +230,10 @@ export const getStyles = (props: ICheckboxStyleProps): ICheckboxStyles => {
         transitionTimingFunction: MS_CHECKBOX_TRANSITION_TIMING,
 
         /* in case the icon is bigger than the box */
-        overflow: 'hidden'
+        overflow: 'hidden',
+        selectors: {
```
Comment (dzearing) [inline packages/office-ui-fabric-react/src/components/Checkbox/Checkbox.styles.ts]: this should be conditionalized to be added only when using indeterminate.
